pkg/sql/execinfrapb/processors_bulk_io.proto

```diff
@@ -59,7 +59,9 @@ message BackfillerSpec {
   optional int64 chunk_size = 5 [(gogoproto.nullable) = false];
 
   // The timestamp to perform index backfill historical scans at.
-  optional util.hlc.Timestamp readAsOf = 7 [(gogoproto.nullable) = false];
+  optional util.hlc.Timestamp writeAsOf = 7 [(gogoproto.nullable) = false];
+
+  optional util.hlc.Timestamp readAsOf = 9 [(gogoproto.nullable) = false];
```

Comment:
For my own education: why is it safe here to reassign tag 7 to `writeAsOf` and give `readAsOf` a new tag? I'm guessing backward and forward compatibility isn't a concern here, but I can't help to naively wonder what happens when we're running a mixed-version cluster? wouldn't you have old protos and new protos both being moved around and interpreted differently on old vs new version nodes? I can't stress the "naive" part of my understanding enough.

Reply:
Backwards compatibility / mixed-version behavior is *absolutely* a concern, which is precisely why I used the old tag (7) for the "new" write-as-of time field and used a new tag (9) for the "old" read-as-time field.

A time we scanned at is the only safe-to-write-at time -- we *must not* tell processors to write at a time unless we also scanned at it.

With this patch, if a new gateway node decides it does not want to re-scan at a new wall-time on a resume, but rather wants the whole flow to write at the old time, it needs to be sure the remote processors -- even if they haven't upgraded yet -- will write at that timestamp. What time do remote processors (pre-patch) write at? the one in field 7 (which they call "readAsOf"). By putting the _write_ time in field 7, an old processor will write at that time. It'll also _read_ at that time which means it might, if that time is far in the past, be under the TTL and the proc might get an error and fail. But that's fine -- you can update node and try again.

Reply:
Thank you for the explanation, this makes sense to me now.